nulib/dc-api-v2
# Implements the OAI-PMH protocol at the '/oai' endpoint.

## Description
Implements the OAI-PMH protocol at the '/oai' endpoint. The endpoint supports both 'GET' and 'POST' requests according to the specification. All responses should return XML documents that validate against the [OAI_PMH schema](http://www.openarchives.org/OAI/2.0/oai_dc.xsd)

**Supports the following verbs:**
- [GetRecord](http://www.openarchives.org/OAI/2.0/openarchivesprotocol.2002-08-06.htm#GetRecord) 
- [Identify](http://www.openarchives.org/OAI/2.0/openarchivesprotocol.2002-08-06.htm#Identify) 
- [ListIdentifiers](http://www.openarchives.org/OAI/2.0/openarchivesprotocol.2002-08-06.htm#ListIdentifers)
- [ListMetadataFormats](http://www.openarchives.org/OAI/2.0/openarchivesprotocol.2002-08-06.htm#ListMetadataFormats)
- [ListRecords](http://www.openarchives.org/OAI/2.0/openarchivesprotocol.2002-08-06.htm#ListRecords) 
- [ListSets](http://www.openarchives.org/OAI/2.0/openarchivesprotocol.2002-08-06.htm#ListSets)

**Tests:**
100% test coverage, check out `test/integration/oai.test.js` to see all the error case handling required by the OAI-PMH specification  

## Steps to test
For "list requests" (e.g. `ListRecords`):
- Make a request using the `verb` and `metadataPrefix` parameters: `http://localhost:3000/oai?verb=ListRecords&metadataPrefix=oai_dc`
- Make subsequent requests including the `resumptionToken` returned in the previous request: `http://localhost:3000/oai?verb=ListRecords&metadataPrefix=oai_dc&resumptionToken=FGluY2x1ZGVfY29udGV4dF91dWlkDXF1ZXJ5QW5kRmV0Y2gBFm1jN3ZCajdnUURpbUhad1hIYnNsQmcAAAAAAAB2DhZXbmtMZVF5Q1JsMi1ScGRsYUlHLUtB`
- Repeat the previous step until the list is complete. You'll know it is complete when you get an XML response with an empty `resumptionToken` element
- If you continue to make requests using that same `resumptionToken` you should get an error response with `badResumptionToken`

For all other verbs, pass in the verb and any required parameters using either `POST` or `GET` requests:
- `curl http://localhost:3000/oai?verb=GetRecord&identifier=c02c149f-1df1-4252-ae93-79031c676e71&metadataPrefix=oai_dc`
- `curl -X POST 'http://localhost:3000/oai' --data 'verb=Identify'`
